students/douglas_klos/session9/examples/object_canvas.py

```diff
@@ -0,0 +1,173 @@
+
+    def add_object(self, draw_object, position="top"):
+        #fixme: maybe overload the inplace addition operator?
+        """
+        Add a new object to the canvas.
+
+        :param: draw_object -- DrawObject to add
+
+        :param position="top": Position to add the object. "top" puts
+                               the object on top of teh other objects.
+                               "bottom" puts them on the bottom of the stack.
+                               A integer puts it in that place in the order
+                               -- 0 is the bottom.
+        """
+        if position == "top":
+            self.draw_objects.append(draw_object)
+        elif position == "bottom":
+            self.draw_objects.insert(0, draw_object)
+        else:
+            self.draw_objects.insert(position, draw_object)
+
+    def render(self, filename):
+        """
+        render the drawing to a file with the given name
+        """
+        image = Image.new('RGBA', self.size, color=self.background)
+        drawer = ImageDraw.Draw(image)
+
+        for do in self.draw_objects:
+            do.draw(drawer)
+        image.save(filename)
+
+
+class DrawObject:
+    """
+    base class for all draw objects
+    """
+
+    def __init__(self, *args, **kwargs):
+        print("in DrawObject __init__", kwargs)
+        # do nothing, but to make super happy
+        super().__init__(*args, **kwargs)
+
+
+class LineObject:
+    """
+    mixin for classes with a line
+    """
+
+    def __init__(self,
+                 line_color='black',
+                 line_width=1,
+                 **kwargs,
+                 ):
+        print("in LineObject __init__", kwargs)
+        super().__init__(**kwargs)
+        self.line_color = line_color
+        self.line_width = line_width
+
+
+class FillObject:
+    """
+    mixin for classes with a fill
+    """
+
+    def __init__(self,
+                 fill_color=None,
+                 **kwargs
+                 ):
+        print("in FillObject __init__", kwargs)
+        self.fill_color = fill_color
+
+
+class PolyLine(DrawObject, LineObject):
+    def __init__(self,
+                 vertices,
+                 **kwargs
+                 ):
+        self.vertices = vertices
+        print("in PolyLine init", kwargs)
+        super().__init__(**kwargs)
+
+    def draw(self, drawer):
+        """
+        draw the object
+
+        :param drawer: PIL.ImageDraw object to draw to
+        """
+        drawer.line(self.vertices, fill=self.line_color, width=self.line_width)
+
+
+class Circle(DrawObject, LineObject, FillObject):
+    def __init__(self, center, diameter, **kwargs):
+        self.center = center
+        self.diameter = diameter
+        super().__init__(**kwargs)
+
+    def draw(self, drawer):
+        """
+        Draw the object
+        :param drawer: PIL.ImageDraw object to draw to
+        """
+        r = self.diameter // 2
+        c = self.center
```

Comment:
no single letter variable if you can.  be descriptive.

Reply:
I didn't write the things in the example folder, they were available for download on the https://uwpce-pythoncert.github.io/PythonCertDevel/exercises/index.html site.

Reply:
haha, sorry.  many times I mix up what belongs to what and start doing PR comments on instructor code :)